File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class TreeInfo:
    def __init__(self, nodes_visited, result):
        self.nodes_visited = nodes_visited
        self.result = result

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        treeinfo = TreeInfo(0, -1)
        self.inorder_traversal(root, k, treeinfo)
        return treeinfo.result

    def inorder_traversal(self, node, k, treeinfo):
        if not node:
            return TreeInfo(0, -1)

        self.inorder_traversal(node.left, k, treeinfo)
        if treeinfo.nodes_visited < k:
            treeinfo.nodes_visited += 1
            treeinfo.result = node.val
            self.inorder_traversal(node.right, k, treeinfo)
        

File: Data Structures & Algorithms/path-sum/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int target_sum) {
        if (root == nullptr) {
            return false;
        }

        if (root->left == nullptr && root->right == nullptr && abs(target_sum - root->val) == 0) {
            return true;
        }

        bool left_subtree = hasPathSum(root->left, target_sum - root->val);
        bool right_subtree = hasPathSum(root->right, target_sum - root->val);

        return left_subtree || right_subtree;

        
    }
};